File: sdks/local-rp/java/src/test/java/community/catalyst/linkkeys/localrp/ExpirationsConformanceTest.java
package community.catalyst.linkkeys.localrp;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.time.Duration;
import java.time.Instant;
import java.time.OffsetDateTime;
import java.util.List;

import org.junit.jupiter.api.Test;

import community.catalyst.linkkeys.localrp.testutil.Fixtures;
import community.catalyst.linkkeys.localrp.testutil.MiniJson.JsonValue;

/** Conformance vectors: {@code expirations.json}. */
class ExpirationsConformanceTest {

    @Test
    void checkExpirationsThresholdsViaSdkWrapper() {
        JsonValue d = Fixtures.load("expirations.json").get("check_expirations");
        String expiresAt = d.get("expires_at").asString();
        List<JsonValue> cases = d.get("cases").asArray();
        assertEquals(11, cases.size());

        // Build an identity whose descriptor expires at exactly `expiresAt`,
        // exercising `LinkKeysLocalRp.checkExpirations` end to end (identity
        // -> descriptor -> threshold logic), not the underlying LocalRp
        // function directly.
        Instant expires = OffsetDateTime.parse(expiresAt).toInstant();
        Instant createdAt = expires.minus(Duration.ofDays(3650));
        Identity.GenerateLocalRpIdentityConfig config =
                new Identity.GenerateLocalRpIdentityConfig("Conformance Test App", createdAt);
        config.lifetime = Duration.between(createdAt, expires);
        Identity.LocalRpKeyMaterial identity = Identity.generateLocalRpIdentity(config);

        for (JsonValue c : cases) {
            Instant now = OffsetDateTime.parse(c.get("now").asString()).toInstant();
            LocalRp.ExpirationStatus status = LinkKeysLocalRp.checkExpirations(identity, now);
            assertEquals(c.get("expected_level").asString(), status.level().wireName(), "now=" + now);
        }
    }

    @Test
    void checkTimestampsSkewBoundariesAreExact() {
        JsonValue d = Fixtures.load("expirations.json").get("check_timestamps");
        String issuedAt = d.get("issued_at").asString();
        String expiresAt = d.get("expires_at").asString();
        long skew = d.get("skew_seconds").asLong();
        List<JsonValue> cases = d.get("cases").asArray();
        assertEquals(4, cases.size());

        for (JsonValue c : cases) {
            Instant now = OffsetDateTime.parse(c.get("now").asString()).toInstant();
            boolean expectedValid = c.get("expected_valid").asBoolean();
            boolean valid;
            try {
                LocalRp.checkTimestamps(issuedAt, expiresAt, now, skew);
                valid = true;
            } catch (LocalRpError e) {
                valid = false;
            }
            assertEquals(expectedValid, valid, "now=" + now);
        }
    }
}

File: sdks/local-rp/kotlin/src/test/kotlin/community/catalyst/linkkeys/localrp/kt/ExpirationsConformanceTest.kt
package community.catalyst.linkkeys.localrp.kt

import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import community.catalyst.linkkeys.localrp.kt.protocol.checkTimestamps
import community.catalyst.linkkeys.localrp.kt.testutil.Fixtures

/** Conformance vectors: `expirations.json`, exercised through this package's own [checkExpirations] / [checkTimestamps]. */
class ExpirationsConformanceTest {

    @Test
    fun checkExpirationsThresholdsViaSdkWrapper() {
        val d = Fixtures.load("expirations.json").get("check_expirations")
        val expiresAtStr = d.get("expires_at").asString()
        val cases = d.get("cases").asArray()
        assertEquals(11, cases.size)

        // Build an identity whose descriptor expires at exactly `expiresAt`,
        // exercising checkExpirations end to end (identity -> descriptor ->
        // threshold logic), not a raw threshold function directly.
        val expires: Instant = OffsetDateTime.parse(expiresAtStr).toInstant()
        val createdAt = expires.minus(Duration.ofDays(3650))
        val identity = generateLocalRpIdentity(
            appName = "Conformance Test App",
            now = createdAt,
            lifetime = Duration.between(createdAt, expires),
        )

        for (c in cases) {
            val now = OffsetDateTime.parse(c.get("now").asString()).toInstant()
            val status = checkExpirations(identity, now)
            assertEquals(c.get("expected_level").asString(), status.level.name.lowercase(), "now=$now")
        }
    }

    @Test
    fun checkTimestampsSkewBoundariesAreExact() {
        val d = Fixtures.load("expirations.json").get("check_timestamps")
        val issuedAt: Instant = OffsetDateTime.parse(d.get("issued_at").asString()).toInstant()
        val expiresAt: Instant = OffsetDateTime.parse(d.get("expires_at").asString()).toInstant()
        val skew = d.get("skew_seconds").asLong()
        val cases = d.get("cases").asArray()
        assertEquals(4, cases.size)

        for (c in cases) {
            val now = OffsetDateTime.parse(c.get("now").asString()).toInstant()
            val expectedValid = c.get("expected_valid").asBoolean()
            val valid = try {
                checkTimestamps(issuedAt, expiresAt, now, skew)
                true
            } catch (e: LocalRpException.Protocol) {
                false
            }
            assertEquals(expectedValid, valid, "now=$now")
        }
    }
}
